crates/red_knot_python_semantic/src/types.rs

```diff
@@ -397,6 +397,16 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Return true if the type is a class or a union of classes.
+    pub fn is_class(&self, db: &'db dyn Db) -> bool {
+        match self {
+            Type::Union(union) => union.elements(db).iter().all(|ty| ty.is_class(db)),
+            Type::Class(_) => true,
+            // / TODO include type[X], once we add that type
+            _ => false,
+        }
+    }
```

Comment:
I'm not sure we'll do the right thing with this for something like

```py
flag = True

if flag:
    class Foo:
        def __class_getitem__(self, x: int) -> str:
            pass
else:
    class Foo:
        pass

Foo[42]
```

Here we want to:
1. Infer a union type for the `Foo` class
2. Emit a diagnostic because not all items in the union support being subscripted
3. Infer `str | Unknown` for the result of the subscription

Reply:
I think that is working as described? I added a test.

Reply:
Yeah, I was debating whether to bring this up earlier, or whether it's adequate to just say "not subscriptable" in this case. I agree the behavior you're describing is probably more ideal. It might require explicitly handling unions as a special case here, and calling the class-getitem resolution method per element of the union? I'd also be fine with a TODO for this so this PR can land.

Reply:
Are you instead thinking of something like this:

```python
flag = True

if flag:
    class Foo:
        def __class_getitem__(self, x: int) -> str:
            pass
else:
    Foo = 1

Foo[42]
```

Where not all members of the union are classes?

Reply:
Huh, your tests are pretty convincing. Thanks for adding them!

Reply:
I added a test for the case I described above, which _does_ resolve to `Unknown` (with a TODO).

Reply:
Ahh, I think I did spot a bug but didn't accurately identify where the bug would manifest? Anyway, thanks again, a TODO for now is fine by me